File: model/src/main/java/io/github/chriszhong/model/organization/event/ContainsEvent.java
package io.github.chriszhong.model.organization.event;

import io.github.chriszhong.message.M;
import io.github.chriszhong.model.organization.entity.Characteristic;
import io.github.chriszhong.model.organization.entity.Role;
import io.github.chriszhong.model.organization.id.UniqueId;
import io.github.chriszhong.model.organization.relation.Contains;

import java.util.Objects;

import javax.inject.Inject;
import javax.validation.constraints.NotNull;

import com.google.inject.assistedinject.Assisted;

/**
 * The {@linkplain ContainsEvent} class indicates that there is an update about a {@linkplain Contains} relation.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public class ContainsEvent extends AbstractEvent {

	private static final long serialVersionUID = 87203168183879944L;
	private final UniqueId<Role> roleId;
	private final UniqueId<Characteristic> characteristicId;
	private final double value;
	private transient Integer hashCode = null;
	private transient String toString = null;

	@Inject
	ContainsEvent(@NotNull @Assisted final EventCategory category, @NotNull @Assisted final Contains contains) {
		super(category);
		roleId = contains.getRole().getId();
		characteristicId = contains.getCharacteristic().getId();
		value = contains.getValue();
	}

	/**
	 * Returns a {@linkplain UniqueId} that represents a {@linkplain Role}.
	 *
	 * @return a {@linkplain UniqueId}.
	 */
	public UniqueId<Role> getRoleId() {
		return roleId;
	}

	/**
	 * Returns a {@linkplain UniqueId} that represents a {@linkplain Characteristic}.
	 *
	 * @return a {@linkplain UniqueId}.
	 */
	public UniqueId<Characteristic> getCharacteristicId() {
		return characteristicId;
	}

	/**
	 * Returns a <code>double</code> value.
	 *
	 * @return a <code>double</code> value
	 */
	public double getValue() {
		return value;
	}

	@Override
	public boolean equals(final Object object) {
		if (object instanceof ContainsEvent) {
			final ContainsEvent event = (ContainsEvent) object;
			return super.equals(event) && getRoleId().equals(event.getRoleId()) && getCharacteristicId().equals(event.getCharacteristicId());
		}
		return false;
	}

	@Override
	public int hashCode() {
		if (hashCode == null) {
			hashCode = Objects.hash(getCategory(), getRoleId(), getCharacteristicId());
		}
		return hashCode;
	}

	@Override
	public String toString() {
		if (toString == null) {
			toString = M.EVENT_WITH_2_IDS_AND_VALUE.get(super.toString(), getRoleId(), getCharacteristicId(), getValue());
		}
		return toString;
	}

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/relation/ContainsEvent.kt
package runtimemodels.chazm.model.relation

import com.google.inject.assistedinject.Assisted
import runtimemodels.chazm.api.entity.Characteristic
import runtimemodels.chazm.api.entity.CharacteristicId
import runtimemodels.chazm.api.entity.Role
import runtimemodels.chazm.api.entity.RoleId
import runtimemodels.chazm.api.id.UniqueId
import runtimemodels.chazm.api.relation.Contains
import runtimemodels.chazm.model.event.AbstractEvent
import runtimemodels.chazm.model.event.EventType
import runtimemodels.chazm.model.message.M
import java.util.*
import javax.inject.Inject

/**
 * The [ContainsEvent] class indicates that there is an update about a [Contains] relation.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
open class ContainsEvent @Inject internal constructor(
    @Assisted category: EventType,
    @Assisted contains: Contains
) : AbstractEvent(category) {
    /**
     * Returns a [UniqueId] that represents a [Role].
     *
     * @return a [UniqueId].
     */
    val roleId: RoleId = contains.role.id
    /**
     * Returns a [UniqueId] that represents a [Characteristic].
     *
     * @return a [UniqueId].
     */
    val characteristicId: CharacteristicId = contains.characteristic.id
    /**
     * Returns a `double` value.
     *
     * @return a `double` value
     */
    val value: Double = contains.value

    override fun equals(other: Any?): Boolean {
        if (other is ContainsEvent) {
            return super.equals(other) && roleId == other.roleId && characteristicId == other.characteristicId
        }
        return false
    }

    override fun hashCode(): Int = Objects.hash(category, roleId, characteristicId)

    override fun toString(): String = M.EVENT_WITH_2_IDS_AND_VALUE[super.toString(), roleId, characteristicId, value]

    companion object {
        private const val serialVersionUID = 87203168183879944L
    }

}
